yapapi/executor/__init__.py

```diff
@@ -408,34 +399,26 @@ async def start_worker(agreement: rest.market.Agreement) -> None:
             emit(events.WorkerFinished(agr_id=agreement.id))
 
         async def worker_starter() -> None:
-            nonlocal agreements_confirmed
             while True:
                 await asyncio.sleep(2)
-                if (
-                    offer_buffer
-                    and len(workers) < self._conf.max_workers
+                await agreements_pool.cycle()
+                if (len(workers) < self._conf.max_workers
                     and work_queue.has_unassigned_items()
                 ):
-                    provider_id, b = random.choice(list(offer_buffer.items()))
-                    del offer_buffer[provider_id]
                     new_task = None
                     try:
-                        agreement = await b.proposal.create_agreement()
-                        provider = (await agreement.details()).provider_view.extract(NodeInfo)
-                        emit(events.AgreementCreated(agr_id=agreement.id, provider_id=provider))
-                        if not await agreement.confirm():
-                            emit(events.AgreementRejected(agr_id=agreement.id))
+                        new_task = await agreements_pool.use_agreement(lambda a: loop.create_task(start_worker(a)))
+                        if new_task is None:
                             continue
-                        emit(events.AgreementConfirmed(agr_id=agreement.id))
-                        agreements_confirmed += 1
-                        new_task = loop.create_task(start_worker(agreement))
                         workers.add(new_task)
                     except CancelledError:
                         raise
                     except Exception as e:
+                        if self._conf.traceback:
+                            traceback.print_exc()
                         if new_task:
                             new_task.cancel()
-                        emit(events.ProposalFailed(prop_id=b.proposal.id, reason=str(e)))
+                        emit(events.ProposalFailed(prop_id="<unknown>", reason=str(e)))
```

Comment:
I guess you can just assume that `agreements_pool.use_agreement()` emits `ProposalFailed` whenever appropriate and skip this line.

Reply:
I wanted to have some way of reporting when use_agreement fails.

Reply:
Your code already emits `ProposalFailed` in `AgreementPool._get_agreement()`. You can also lift the emitting out of `_get_agreement()` to `use_agreement()`, if that would match the original code better. 

I'd like to avoid emitting `ProposalFailed` twice (once in `AgreementPool` and another one here in `Executor._submit()`). Those events are not only for logging, there may exist some consumers listening to them and relying on `prop_id` being set to a correct value. For logging, this `logger.Logger` instance can be used:
https://github.com/golemfactory/yapapi/blob/f83c74d357289c4a003775c8d2e39e973cc4288d/yapapi/executor/__init__.py#L55